barretenberg/cpp/src/barretenberg/common/thread_utils.hpp

```diff
@@ -0,0 +1,22 @@
+#include "thread.hpp"
+
+namespace barretenberg::thread_utils {
+
+const size_t DEFAULT_MIN_ITERS_PER_THREAD = 1 << 4;
+
+/**
+ * @brief calculates number of threads to create based on minimum iterations per thread
+ * @details Finds the number of cpus with get_num_cpus(), and calculates `desired_num_threads`
+ * Returns the min of `desired_num_threads` and `max_num_theads`.
+ * Note that it will not calculate a power of 2 necessarily, use `calculate_num_threads_pow2` instead
+ *
+ * @param num_iterations
+ * @param min_iterations_per_thread
+ * @return size_t
+ */
+size_t calculate_num_threads(size_t num_iterations, size_t min_iterations_per_thread = DEFAULT_MIN_ITERS_PER_THREAD);
+
+size_t calculate_num_threads_pow2(size_t num_iterations,
```

Comment:
should I have the same comment here and in the cpp file?

Reply:
I'm don't think we have clear consensus on this (and I don't think Doxygen cares). It's not very satisfying but I think this is something that has to be decided on a case by case basis. If the comments include lots of discussion of implementation details, it makes sense to attach them to the implementation. If they're just `@brief` and some interface details, the header is probably fine.

Reply:
Makes sense. Might just copy paste them in this case